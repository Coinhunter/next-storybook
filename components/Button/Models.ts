export enum ButtonTypes {
  BUTTON = 'button',
  SUBMIT = 'submit',
  RESET = 'reset',
}

export enum ButtonUIType {
  OLDPRIMARY = 'oldprimary',
  PRIMARY = 'primary',
  SECONDARY = 'secondary',
  DISABLED = 'disabled',
}

export enum ButtonSize {
  SMALL = 'small',
  MEDIUM = 'medium',
}
