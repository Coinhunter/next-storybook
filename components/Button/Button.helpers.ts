import classNames from 'classnames';
import { ButtonUIType } from './Models';

import styles from './Button.module.css';
import { ButtonSize } from './Models';

export const getClassNamesFromButtonUIType = (buttonUIType: ButtonUIType): string => {
  switch (buttonUIType) {
    case ButtonUIType.OLDPRIMARY:
      return classNames(styles.Button, styles['Button-oldPrimary']);
    case ButtonUIType.PRIMARY:
      return classNames(styles.Button, styles['Button-primary']);
    case ButtonUIType.SECONDARY:
      return classNames(styles.Button, styles['Button-secondary']);
    case ButtonUIType.DISABLED:
      return classNames(styles.Button, styles['Button-disabled']);
    default:
      return classNames(styles.Button, styles['Button-primary']);
  }
};

export const getSmallButtonClassNamesFromButtonSize = (buttonSize: ButtonSize): string => {
  switch (buttonSize) {
    case ButtonSize.SMALL:
      return classNames(styles['IconButton-small']);
    case ButtonSize.MEDIUM:
      return classNames(styles['IconButton-medium']);
    default:
      return classNames(styles['IconButton-medium']);
  }
};
