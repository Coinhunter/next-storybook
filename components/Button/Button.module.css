/** Basic configuration for a button **/
.Button {
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;

  white-space: nowrap;
  cursor: pointer;
  outline: none;
  border: none;
  border-radius: var(--radius-button-dsm);

  padding-left: var(--space-s);
  padding-right: var(--space-s);

  font-weight: bold;

  height: 46px;
}
.Button:focus:after {
  position: absolute;
  content: '';
  top: -4px;
  bottom: -4px;
  right: -4px;
  left: -4px;
  border: 2px solid var(--color-primary-300);
  border-radius: 10px;
}

/** Line height needs to be reset for icon to display properly **/
.Button-loading {
  line-height: 0;
}

/** The styling for the current web button is represented as oldPrimary **/
.Button-oldPrimary {
  color: var(--c_background);
  background-color: var(--c_color_alpha);
  border-radius: var(--d_radius_large);
}
.Button-oldPrimary:focus:not(:disabled) {
  color: var(--c_background);
  background-color: var(--c_color_alpha);
}
.Button-oldPrimary:focus:after {
  color: var(--c_background);
  top: -8px;
  bottom: -8px;
  right: -8px;
  left: -8px;
  border: 2px solid var(--c_color_alpha);
  border-radius: 22px;
}
.Button-oldPrimary:hover:not(:disabled) {
  color: var(--c_background);
  background-color: var(--c_color_alpha_dark);
}
.Button-oldPrimary:active:not(:disabled) {
  color: var(--c_background);
  background-color: var(--c_color_alpha_dark);
}

/** Design system button Primary **/
.Button-primary {
  color: var(--c_background);
  background-color: var(--color-primary-400);
}
.Button-primary:focus:not(:disabled) {
  color: var(--c_background);
  background-color: var(--color-primary-400);
}
.Button-primary:hover:not(:disabled) {
  color: var(--c_background);
  background-color: var(--color-primary-500);
}
.Button-primary:active:not(:disabled) {
  color: var(--c_background);
  background-color: var(--color-primary-500);
}

/** Design system button Secondary **/
.Button-secondary {
  color: var(--color-neutral-800);
  background-color: var(--color-neutral-200);
}
.Button-secondary:focus:not(:disabled) {
  color: var(--color-neutral-800);
  background-color: var(--color-neutral-200);
}
.Button-secondary:hover:not(:disabled) {
  color: var(--color-neutral-800);
  background-color: var(--color-neutral-300);
}
.Button-secondary:active:not(:disabled) {
  color: var(--color-neutral-800);
  background-color: var(--color-neutral-400);
}

/** Design system button Disabled **/
.Button-disabled {
  opacity: 0.7;
  cursor: not-allowed;
  background: var(--color-neutral-200);
  color: var(--color-neutral-300);
}
.Button-disabled:focus {
  background: var(--color-neutral-200);
  color: var(--color-neutral-300);
}
.Button-disabled:hover {
  background: var(--color-neutral-200);
  color: var(--color-neutral-300);
}
.Button-disabled:active {
  background: var(--color-neutral-200);
  color: var(--color-neutral-300);
}

/** Fullwidth-class is applied when the fullwidth prop is given **/
.Button-fullWidth {
  width: 100%;
}

/** Design system link (href) style as button **/
.Button-link {
  display: inline-flex;
  text-decoration: none;
}

/** Small button class, change paddings and height. **/
.Button-small {
  padding-left: var(--space-xxs);
  padding-right: var(--space-xxs);
  height: 32px;
}

.IconButton {
  padding: 0;
  border-radius: var(--radius-button-dsm);
}
.IconButton:focus:after {
  border-radius: 10px;
  top: -4px;
  bottom: -4px;
  right: -4px;
  left: -4px;
  border: 2px solid var(--color-primary-300);
  border-radius: 10px;
}

.IconButton-medium {
  height: 46px;
  width: 46px;
}

.IconButton-small {
  height: 32px;
  width: 32px;
}

/* TODO: Figure this out! */
.c-buttonPlain {
  background-color: transparent;
  border: none;
  margin: 0;
  padding: 0;
  display: flex;
  outline: none;
  cursor: pointer;
  border: 2px solid transparent;
}
.c-buttonPlain:focus {
  outline: none;
  border: 2px solid var(--c_color_alpha);
  border-radius: var(--radius-button-dsm);
}
