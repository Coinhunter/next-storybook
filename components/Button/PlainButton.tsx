import React from 'react';

import styles from './Button.module.css';

interface Props {
  id: string;
  children: any;
  onClick?: (e: React.MouseEvent<HTMLButtonElement>) => void;
  onEnter: (e: React.KeyboardEvent<HTMLButtonElement>) => void;
  onKeyDown?: (event: React.KeyboardEvent<HTMLButtonElement>) => void;
  onBlur?: React.FocusEventHandler<HTMLButtonElement>;
  onFocus?: React.FocusEventHandler<HTMLButtonElement>;
}

const PlainButton = (props: Props): React.ReactElement => {
  const { children, onClick, onEnter, id, onKeyDown, onBlur, onFocus } = props;

  const onKeyDownInternal = (event: React.KeyboardEvent<HTMLButtonElement>): void => {
    event.key === 'Enter' && onEnter(event);
    onKeyDown && onKeyDown(event);
  };

  const onBlurInternal = (event: React.FocusEvent<HTMLButtonElement>) => {
    onBlur && onBlur(event);
  };

  const onFocusInternal = (event: React.FocusEvent<HTMLButtonElement>) => {
    onFocus && onFocus(event);
  };

  return (
    <button
      id={id}
      className={styles['c-buttonPlain']}
      onClick={onClick}
      onKeyDown={onKeyDownInternal}
      onBlur={onBlurInternal}
      onFocus={onFocusInternal}>
      {children}
    </button>
  );
};

PlainButton.displayName = 'PlainButton';

export { PlainButton };
