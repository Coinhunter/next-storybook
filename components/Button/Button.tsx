import React, { HTMLAttributes } from 'react';
import classNames from 'classnames';

import styles from './Button.module.css';
import { Spinner } from '../Spinner';
import { ButtonTypes, ButtonUIType, ButtonSize } from './Models';
import { getClassNamesFromButtonUIType } from './Button.helpers';

interface Props extends HTMLAttributes<HTMLAnchorElement | HTMLButtonElement> {
  /** id of the button */
  id: string;
  /** Type: BUTTON | SUBMIT | RESET | LINK*/
  type?: ButtonTypes;
  /** ButtonUIType: OLDPRIMARY | PRIMARY | SECONDARY | DISABLED */
  buttonUIType?: ButtonUIType;
  /** ButtonSize: SMALL | NORMAL - Does not work for OLDPRIMARY type.  */
  buttonSize?: ButtonSize;
  /** href: string */
  href?: string;
  /** children:  */
  children?: React.ReactNode;
  /** Listen to focus events */
  onFocus?: (e: React.FocusEvent<HTMLAnchorElement | HTMLButtonElement>) => void;
  /** Listen to blur events */
  onBlur?: (e: React.FocusEvent<HTMLAnchorElement | HTMLButtonElement>) => void;
  /** Listen to click events */
  onClick?: (e: React.MouseEvent<HTMLAnchorElement | HTMLButtonElement>) => void;
  /** Listen to onKeyDown events */
  onKeyDown?: (e: React.KeyboardEvent<HTMLAnchorElement | HTMLButtonElement>) => void;
  /** aria-label: string */
  ariaLabel?: string;
  /** if button renders as a anchor tag we need to be able to add rel */
  rel?: string;
  /** target for anchor tags */
  target?: string;
  /** id used for automated tests */
  dataTestId?: string;
  /** Full width */
  fullWidth?: boolean;
  /** loading state on button */
  isLoading?: boolean;
  /** optional css class name for button */
  className?: string;
  /** CSS class for children wrapper */
  childrenClassName?: string;
}

const Button = (props: Props) => {
  const { id, dataTestId = id, ariaLabel } = props;
  const {
    type = ButtonTypes.BUTTON,
    buttonSize = ButtonSize.MEDIUM,
    buttonUIType = ButtonUIType.OLDPRIMARY,
    children,
    fullWidth = false,
    isLoading = false,
  } = props;
  const {
    onFocus,
    onBlur,
    onClick,
    onKeyDown,
    rel,
    target,
    href,
    className: buttonClassName,
    childrenClassName,
  } = props;

  // If we got a href value we should render a link.
  if (href) {
    return (
      <a
        id={id}
        href={href}
        aria-label={ariaLabel}
        target={target}
        rel={rel}
        onClick={(event: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => {
          if (buttonUIType === ButtonUIType.DISABLED) {
            event.stopPropagation();
            event.preventDefault();
          }
        }}
        data-test-id={dataTestId}
        className={classNames(
          getClassNamesFromButtonUIType(buttonUIType),
          styles['Button-link'],
          fullWidth && styles['Button-fullWidth'],
          isLoading && styles['Button-loading'],
          buttonSize === ButtonSize.SMALL && styles['Button-small'],
          buttonClassName,
        )}>
        <span className={classNames(childrenClassName)}>{isLoading ? <Spinner size={'24px'} /> : <>{children}</>}</span>
      </a>
    );
  }

  return (
    <button
      id={id}
      type={type}
      onFocus={onFocus}
      onBlur={onBlur}
      onClick={onClick}
      onKeyDown={onKeyDown}
      aria-label={ariaLabel}
      data-test-id={dataTestId}
      className={classNames(
        getClassNamesFromButtonUIType(buttonUIType),
        fullWidth && styles['Button-fullWidth'],
        isLoading && styles['Button-loading'],
        buttonSize === ButtonSize.SMALL && styles['Button-small'],
        buttonClassName,
      )}
      disabled={buttonUIType === ButtonUIType.DISABLED}>
      <span className={classNames(childrenClassName)}>{isLoading ? <Spinner size={'24px'} /> : <>{children}</>}</span>
    </button>
  );
};

Button.displayName = 'Button';

export { Button };

export default Button;
