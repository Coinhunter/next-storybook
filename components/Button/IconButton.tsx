import React, { HTMLAttributes } from 'react';
import classNames from 'classnames';

import { ButtonTypes, ButtonUIType, ButtonSize } from './Models';
import { getClassNamesFromButtonUIType, getSmallButtonClassNamesFromButtonSize } from './Button.helpers';
import { Icon, IconName, IconSize } from '../Icon/Icon';

import styles from './Button.module.css';

interface Props extends HTMLAttributes<HTMLAnchorElement | HTMLButtonElement> {
  /** id of the button */
  id: string;
  /** Type: BUTTON | SUBMIT | RESET | LINK*/
  type?: ButtonTypes;
  /** ButtonUIType: OLDPRIMARY | PRIMARY | SECONDARY | DISABLED */
  buttonUIType?: ButtonUIType;
  /** ButtonSize: SMALL | NORMAL - Does not work for OLDPRIMARY type.  */
  buttonSize?: ButtonSize;
  /** Listen to focus events */
  onFocus?: (e: React.FocusEvent<HTMLAnchorElement | HTMLButtonElement>) => void;
  /** Listen to blur events */
  onBlur?: (e: React.FocusEvent<HTMLButtonElement>) => void;
  /** Listen to click events */
  onClick?: (e: React.MouseEvent<HTMLAnchorElement | HTMLButtonElement>) => void;
  /** Listen to onKeyDown events */
  onKeyDown?: (e: React.KeyboardEvent<HTMLAnchorElement | HTMLButtonElement>) => void;
  /** aria-label: string */
  ariaLabel?: string;
  /** id used for automated tests */
  dataTestId?: string;
  /** Icon */
  iconName: IconName;
}

const IconButton = (props: Props) => {
  const {
    id,
    dataTestId = id,
    ariaLabel,
    type = ButtonTypes.BUTTON,
    buttonSize = ButtonSize.MEDIUM,
    buttonUIType = ButtonUIType.OLDPRIMARY,
    iconName,
  } = props;
  const { onFocus, onBlur, onClick, onKeyDown } = props;

  return (
    <button
      id={id}
      type={type}
      onFocus={onFocus}
      onBlur={onBlur}
      onClick={onClick}
      onKeyDown={onKeyDown}
      aria-label={ariaLabel}
      data-test-id={dataTestId}
      className={classNames(
        styles['IconButton'],
        getClassNamesFromButtonUIType(buttonUIType),
        getSmallButtonClassNamesFromButtonSize(buttonSize),
      )}
      disabled={buttonUIType === ButtonUIType.DISABLED}>
      <Icon iconName={iconName} size={buttonSize === ButtonSize.MEDIUM ? IconSize.MEDIUM : IconSize.SMALL} />
    </button>
  );
};

IconButton.displayName = 'IconButton';

export { IconButton };

export default IconButton;
