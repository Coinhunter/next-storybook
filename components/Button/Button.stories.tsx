import React from "react";
import { withKnobs, boolean, radios, text } from "@storybook/addon-knobs";

import { Button as ButtonComponent } from "./Button";
import { IconButton as IconButtonComponent } from "./IconButton";
import { ButtonUIType, ButtonSize } from "./Models";
import { Container } from "../Container";
import { IconName } from "../Icon";

const defaultStory = {
  title: "Button",
  component: ButtonComponent,
  decorators: [withKnobs],
};

const typeOptions = {
  oldprimary: ButtonUIType.OLDPRIMARY,
  primary: ButtonUIType.PRIMARY,
  secondary: ButtonUIType.SECONDARY,
  disabled: ButtonUIType.DISABLED,
};

const sizeOptions = {
  medium: ButtonSize.MEDIUM,
  small: ButtonSize.SMALL,
};

export const Button = () => {
  return (
    <Container>
      <ButtonComponent
        id="primary-button"
        onClick={() => console.log("Clicked button!")}
        buttonUIType={radios(
          "ButtonType",
          typeOptions,
          ButtonUIType.OLDPRIMARY
        )}
        isLoading={boolean("Loading", true)}
        fullWidth={boolean("Fullwidth", false)}
        buttonSize={radios("ButtonSize", sizeOptions, ButtonSize.MEDIUM)}
        href={text("HREF", undefined)}
      >
        I'm a fabulous button
      </ButtonComponent>
    </Container>
  );
};

export const NoPropButton = () => (
  <Container>
    <ButtonComponent
      id="primary-button"
      onClick={() => console.log("Clicked button!")}
    >
      I'm a fabulous button
    </ButtonComponent>
  </Container>
);

export const IconButton = () => (
  <Container>
    <IconButtonComponent
      id="icon-button"
      iconName={IconName.CHEVRON_RIGHT}
      onClick={() => console.log("Clicked button!")}
      buttonUIType={radios("ButtonType", typeOptions, ButtonUIType.OLDPRIMARY)}
      buttonSize={radios("ButtonSize", sizeOptions, ButtonSize.MEDIUM)}
    />
  </Container>
);

export default defaultStory;
