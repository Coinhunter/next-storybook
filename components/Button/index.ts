import { Button } from './Button';
import { IconButton } from './IconButton';
import { ButtonSize, ButtonTypes, ButtonUIType } from './Models';

export { Button, ButtonSize, ButtonTypes, ButtonUIType, IconButton };
