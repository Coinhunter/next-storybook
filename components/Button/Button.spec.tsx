import React from "react";
import renderer from "react-test-renderer";

import { Button } from "./Button";

describe("[Component]: Button", () => {
  it("should match snapshot", () => {
    const component = renderer.create(<Button id="first-button" />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
