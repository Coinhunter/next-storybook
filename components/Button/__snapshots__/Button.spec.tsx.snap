// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component]: Button should match snapshot 1`] = `
<button
  className="Button Button-oldPrimary"
  data-test-id="first-button"
  disabled={false}
  id="first-button"
  type="button"
>
  <span
    className=""
  />
</button>
`;
